app/src/main/java/com/conupods/OutdoorMaps/View/Directions/NavigationActivity.java

```diff
@@ -0,0 +1,217 @@
+        Intent modeSelectIntent = getIntent();
+        unpackIntent(modeSelectIntent);
+
+        // Initialize the GeoAPI context
+        GAC = new GeoApiContext.Builder()
+                .apiKey(getString(R.string.Google_API_Key))
+                .build();
+
+        // Compute the directions, Update the view, and add the polylines
+        computeDirections(mOriginCoordinates, mDestinationCoordinates, mMode);
+
+        // Create the recycler view
+        layoutBottomSheet = findViewById(R.id.bottom_sheet);
+        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.routes_recycler_view);
+        stepsList = new ArrayList<>();
+        mAdapter = new RouteAdapter(stepsList);
+        RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(getApplicationContext());
+        recyclerView.setLayoutManager(mLayoutManager);
+        recyclerView.setItemAnimator(new DefaultItemAnimator());
+        recyclerView.setAdapter(mAdapter);
+    }
+
+    @Override
+    public void onBackPressed() {
+        Intent intent = new Intent(this, ModeSelectActivity.class);
+        loadLocationsIntoIntent(intent);
+        stepsList.clear();
+        startActivity(intent);
+        finish();
+    }
+
+    public void unpackIntent(Intent intent) {
+        mOriginCoordinates = intent.getParcelableExtra("fromCoordinates");
+        mOriginLongName = intent.getStringExtra("fromLongName");
+        mOriginCode = intent.getStringExtra("fromCode");
+
+        mDestinationCoordinates = intent.getParcelableExtra("toCoordinates");
+        mDestinationLongName = intent.getStringExtra("toLongName");
+        mDestinationCode = intent.getStringExtra("toCode");
+
+        mMode = (TravelMode) intent.getSerializableExtra("mode");
+    }
+
+    private void loadLocationsIntoIntent(Intent intent) {
+        intent.putExtra("fromCoordinates", mOriginCoordinates);
+        intent.putExtra("fromCode", mOriginCode);
+        intent.putExtra("fromLongName", mOriginLongName);
+
+        intent.putExtra("toCoordinates", mDestinationCoordinates);
+        intent.putExtra("toCode", mDestinationCode);
+        intent.putExtra("toLongName", mDestinationLongName);
+    }
+
+    // Sends Directions API request
+    // Calls function to update the view elements on success
+    public void computeDirections(LatLng origin, LatLng destination, TravelMode mode) {
+
+        DirectionsApiRequest directions = new DirectionsApiRequest(GAC);
+
+        directions.origin(new com.google.maps.model.LatLng(origin.latitude, origin.longitude));
+        directions.destination(new com.google.maps.model.LatLng(destination.latitude, destination.longitude));
+        directions.mode(mode);
+
+        directions.setCallback(new PendingResult.Callback<DirectionsResult>() {
+
+            @Override
+            public void onResult(DirectionsResult result) {
+
+                if (result == null || result.routes.length == 0 || result.routes[0].legs.length == 0) {
```

Comment:
this may need to be embedded if statements, its vulnerable to a null pointer exception

Reply:
changed